### PR TITLE
cardano-node:  implement --shutdown-on-block-synced

### PR DESCRIPTION
**Pull request by deepfire**
This:

1. adds `--shutdown-on-block-synced BLOCKNO` option, in parallel to `--shutdown-on-slot-synced SLOTNO`
2. integrates it into the workbench

### CODE REVIEW

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs [hunk 1]
```diff
@@ -51,14 +50,18 @@ data ShutdownTrace
   -- ^ Received shutdown request but found unexpected input in --shutdown-ipc FD:
   | RequestingShutdown Text
   -- ^ Ringing the node shutdown doorbell for reason
-  | ShutdownArmedAtSlot SlotNo
-  -- ^ Will terminate upon reaching maxSlot
+  | ShutdownArmedAtSlotBlock (Maybe SlotNo) (Maybe BlockNo)
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs:53]: @Jimbo4350, good point -- I was considering that as well!

[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs]: @Jimbo4350, done.

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs [hunk 2]
```diff
@@ -92,28 +129,40 @@ withShutdownHandling ShutdownConfig{scIPC = Just fd} tr action = do
...
+      traceWith tr (ShutdownArmedAt lim)
+      spawnLimitTerminator lim
  where
-  spawnSlotLimitTerminator :: SlotNo -> IO ()
-  spawnSlotLimitTerminator maxSlot =
+  spawnLimitTerminator :: SlotOrBlock Identity -> IO ()
+  spawnLimitTerminator limit =
     void $ forkLinkedWatcher registry "slotLimitTerminator" Watcher {
-        wFingerprint = id
+        wFingerprint = identity
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Handlers/Shutdown.hs:148]: Why the change to `identity`?

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs [hunk 3]
```diff
@@ -58,7 +58,7 @@ import           Cardano.BM.Stats
 import           Cardano.BM.Tracing (HasPrivacyAnnotation (..), HasSeverityAnnotation (..),
                    Severity (..), ToObject (..), Tracer (..), TracingVerbosity (..),
                    Transformable (..))
-import           Cardano.Slotting.Block (BlockNo (..))
+import           Cardano.Node.Handlers.Shutdown ()
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs:61]: Why remove an orphan instance from here? We should leave orphan instances  here.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
A couple of comments

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs [hunk 4]
```diff
@@ -20,27 +27,50 @@ module Cardano.Node.Handlers.Shutdown
...
 import           "contra-tracer" Control.Tracer
+import           Ouroboros.Consensus.Block (Header)
 import qualified Ouroboros.Consensus.Storage.ChainDB as ChainDB
 import           Ouroboros.Consensus.Util.ResourceRegistry (ResourceRegistry)
 import           Ouroboros.Consensus.Util.STM (Watcher (..), forkLinkedWatcher)
-import           Ouroboros.Network.Block (MaxSlotNo (..), SlotNo, pointSlot)
+import           Ouroboros.Network.Block (HasHeader, BlockNo (..), SlotNo (..), pointSlot)
+
+
+data SlotOrBlock f
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs:49]: Also the name `SlotOrBlock` doesn't really indicate immediately what it's going to be used for. Why not `ShutdownOnSlotOrBlock`? Or something similar.

**Comment by Jimbo4350** [comment]
I think you are using a different formatter to cardano-node. We use stylish-haskell.

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs (hunk 2 above)
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs]: @Jimbo4350, sorry, `Cardano.Prelude` import confused me, undone.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs (hunk 3 above)
Comment (deepfire) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs:61]: @Jimbo4350, because it turned out there's _another_ pre-existing orphan instance elsewhere -- so I just reused that.

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs]: @Jimbo4350, done.

File: cardano-node/src/Cardano/Node/Parsers.hs [hunk 5]
```diff
@@ -93,7 +91,7 @@ nodeRunParser = do
              }
            , pncValidateDB = validate
            , pncShutdownConfig =
-               Last . Just $ ShutdownConfig (getLast shutdownIPC) (getLast shutdownOnSlot)
+               Last . Just $ ShutdownConfig (getLast shutdownIPC) (join $ getLast shutdownOnLimit)
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Parsers.hs]: ..and thank you for an actionable suggestion, too!

**Comment by deepfire** [comment]
Ok, I need to integrate `stylish-haskell` into `emacs`.

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs:151]: Why did you have to make this change? You shouldn't have to.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs:61]: My apologies. I thought you moved it to another module.

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cardano-node/src/Cardano/Node/Handlers/Shutdown.hs (hunk 2 above)
Comment (deepfire) [reply cardano-node/src/Cardano/Node/Handlers/Shutdown.hs:151]: I needed `Generic`, and so instead of adding `GHC.Generics`, I switched the prelude, which allowed dumping a bunch of imports: https://github.com/input-output-hk/cardano-node/pull/3932/commits/2c020c824970a9a8b31928dfc894f282630fe0e7#diff-9a71b006f6ddf2ab12e67e4d7e2951abf353bc8841f8a2638a1336a76b2ae43bL23-L32

**Comment by deepfire** [comment]
All checks passed, node-side review green -- merging.
